In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv("../data/train.csv")

df.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [3]:
y = df["SalePrice"]
X = df.drop("SalePrice", axis=1)

In [4]:
df.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [5]:
# Remove ID column
X = X.drop("Id", axis=1)

# Feature Engineering

# House age
X["HouseAge"] = df["YrSold"] - df["YearBuilt"]

# Remodel age
X["RemodelAge"] = df["YrSold"] - df["YearRemodAdd"]

# Total bathrooms
X["TotalBathrooms"] = (
    df["FullBath"]
    + 0.5 * df["HalfBath"]
    + df["BsmtFullBath"]
    + 0.5 * df["BsmtHalfBath"]
)

# Total house square footage
X["TotalSF"] = (
    df["TotalBsmtSF"]
    + df["1stFlrSF"]
    + df["2ndFlrSF"]
)

# Total porch area
X["TotalPorchSF"] = (
    df["OpenPorchSF"]
    + df["EnclosedPorch"]
    + df["3SsnPorch"]
    + df["ScreenPorch"]
    + df["WoodDeckSF"]
)

# Garage age
X["GarageAge"] = df["YrSold"] - df["GarageYrBlt"]

# Total living area
X["TotalLivingArea"] = df["GrLivArea"] + df["TotalBsmtSF"]

In [6]:
X.shape

(1460, 86)

In [7]:
df.shape

(1460, 81)

In [8]:
X.head()

,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,LotConfig,...,YrSold,SaleType,SaleCondition,HouseAge,RemodelAge,TotalBathrooms,TotalSF,TotalPorchSF,GarageAge,TotalLivingArea
0,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,Inside,...,2008,WD,Normal,5,5,3.5,2566,61,5.0,2566
1,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,FR2,...,2007,WD,Normal,31,31,2.5,2524,298,31.0,2524
2,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,Inside,...,2008,WD,Normal,7,6,3.5,2706,42,7.0,2706
3,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,Corner,...,2006,WD,Abnorml,91,36,2.0,2473,307,8.0,2473
4,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,FR2,...,2008,WD,Normal,8,8,3.5,3343,276,8.0,3343


In [9]:
X.info()

<class 'pandas.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 86 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   MSSubClass       1460 non-null   int64  
 1   MSZoning         1460 non-null   str    
 2   LotFrontage      1201 non-null   float64
 3   LotArea          1460 non-null   int64  
 4   Street           1460 non-null   str    
 5   Alley            91 non-null     str    
 6   LotShape         1460 non-null   str    
 7   LandContour      1460 non-null   str    
 8   Utilities        1460 non-null   str    
 9   LotConfig        1460 non-null   str    
 10  LandSlope        1460 non-null   str    
 11  Neighborhood     1460 non-null   str    
 12  Condition1       1460 non-null   str    
 13  Condition2       1460 non-null   str    
 14  BldgType         1460 non-null   str    
 15  HouseStyle       1460 non-null   str    
 16  OverallQual      1460 non-null   int64  
 17  OverallCond      1460 non

In [10]:
X = X.drop(columns=["Alley","PoolQC","Fence","MiscFeature"])

In [11]:
cat_cols = X.select_dtypes(include="object").columns

X[cat_cols] = X[cat_cols].fillna("None")

C:\Users\Muhammad Iqbal\AppData\Local\Temp\ipykernel_4456\1352129243.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = X.select_dtypes(include="object").columns


In [12]:
num_cols = X.select_dtypes(include=["int64","float64"]).columns

X[num_cols] = X[num_cols].fillna(X[num_cols].median())

In [13]:
X = pd.get_dummies(X)

In [14]:
X.shape

(1460, 293)

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [16]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [17]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

lr_model = LinearRegression()

# train
lr_model.fit(X_train, y_train)

# predict
lr_preds = lr_model.predict(X_test)

# evaluate
lr_mse = mean_squared_error(y_test, lr_preds)
lr_rmse = np.sqrt(lr_mse)
lr_r2 = r2_score(y_test, lr_preds)

print("Linear Regression MSE:", lr_mse)
print("Linear Regression RMSE:", lr_rmse)
print("Linear Regression R2:", lr_r2)

Linear Regression MSE: 811453472.8084117
Linear Regression RMSE: 28486.022411147747
Linear Regression R2: 0.8942087401643913


In [18]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

dt_model = DecisionTreeRegressor(random_state=42)

# train
dt_model.fit(X_train, y_train)

# predict
dt_preds = dt_model.predict(X_test)

# evaluate
dt_mse = mean_squared_error(y_test, dt_preds)
dt_rmse = np.sqrt(dt_mse)
dt_r2 = r2_score(y_test, dt_preds)

print("Decision Tree MSE:", dt_mse)
print("Decision Tree RMSE:", dt_rmse)
print("Decision Tree R2:", dt_r2)

Decision Tree MSE: 1634580940.9075344
Decision Tree RMSE: 40429.951037659375
Decision Tree R2: 0.7868955117742026


In [19]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

rf_model = RandomForestRegressor(random_state=42)

# train
rf_model.fit(X_train, y_train)

# predict
rf_preds = rf_model.predict(X_test)

# evaluate
rf_mse = mean_squared_error(y_test, rf_preds)
rf_rmse = np.sqrt(rf_mse)
rf_r2 = r2_score(y_test, rf_preds)

print("Random Forest MSE:", rf_mse)
print("Random Forest RMSE:", rf_rmse)
print("Random Forest R2:", rf_r2)

Random Forest MSE: 859673750.8602198
Random Forest RMSE: 29320.19356791868
Random Forest R2: 0.8879221394711085


In [20]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

gb_model = GradientBoostingRegressor(random_state=42)

# train
gb_model.fit(X_train, y_train)

# predict
gb_preds = gb_model.predict(X_test)

# evaluate
gb_mse = mean_squared_error(y_test, gb_preds)
gb_rmse = np.sqrt(gb_mse)
gb_r2 = r2_score(y_test, gb_preds)

print("Gradient Boosting MSE:", gb_mse)
print("Gradient Boosting RMSE:", gb_rmse)
print("Gradient Boosting R2:", gb_r2)

Gradient Boosting MSE: 683371327.0585375
Gradient Boosting RMSE: 26141.371942928654
Gradient Boosting R2: 0.9109071363329743


In [21]:
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

svr_model = SVR()

# train
svr_model.fit(X_train, y_train)

# predict
svr_preds = svr_model.predict(X_test)

# evaluate
svr_mse = mean_squared_error(y_test, svr_preds)
svr_rmse = np.sqrt(svr_mse)
svr_r2 = r2_score(y_test, svr_preds)

print("SVR MSE:", svr_mse)
print("SVR RMSE:", svr_rmse)
print("SVR R2:", svr_r2)

SVR MSE: 7858268812.371254
SVR RMSE: 88646.87705932597
SVR R2: -0.024502557010942505


In [22]:
import pandas as pd

results = {
    "Linear Regression": [lr_mse, lr_rmse, lr_r2],
    "Decision Tree": [dt_mse, dt_rmse, dt_r2],
    "Random Forest": [rf_mse, rf_rmse, rf_r2],
    "Gradient Boosting": [gb_mse, gb_rmse, gb_r2],
    "SVR": [svr_mse, svr_rmse, svr_r2]
}

results_df = pd.DataFrame(results, index=["MSE", "RMSE", "R2"]).T

print(results_df)

                            MSE          RMSE        R2
Linear Regression  8.114535e+08  28486.022411  0.894209
Decision Tree      1.634581e+09  40429.951038  0.786896
Random Forest      8.596738e+08  29320.193568  0.887922
Gradient Boosting  6.833713e+08  26141.371943  0.910907
SVR                7.858269e+09  88646.877059 -0.024503


# Feature Engineering Experiment Report

## 1. Objective
The goal of this experiment was to improve model performance by creating new meaningful features and evaluating how feature engineering affects regression models. Feature engineering helps models better capture relationships in the data by transforming or combining existing features into more informative representations.

---

## 2. Dataset
The dataset contains **1460 housing records** with multiple features describing properties such as size, location, condition, and construction details.

Target Variable:
- **SalePrice**

The objective of the models is to predict house prices using these features.

---

## 3. Feature Engineering

Several new features were created using domain knowledge about housing properties.

### House Age
HouseAge = YrSold − YearBuilt

This represents how old the house was when sold. Older houses typically have lower value compared to newer houses.

---

### Remodel Age
RemodelAge = YrSold − YearRemodAdd

This represents how long it has been since the house was remodeled. Recently renovated houses generally have higher market value.

---

### Total Bathrooms
Bathrooms were combined into a single feature:

TotalBathrooms = FullBath + 0.5 × HalfBath + BsmtFullBath + 0.5 × BsmtHalfBath

Half bathrooms were counted as 0.5 since they are less valuable than full bathrooms.

---

### Total Square Footage
TotalSF = TotalBsmtSF + 1stFlrSF + 2ndFlrSF

This feature represents the overall size of the house including basement and upper floors.

---

### Total Porch Area
TotalPorchSF = OpenPorchSF + EnclosedPorch + 3SsnPorch + ScreenPorch + WoodDeckSF

Outdoor space contributes to property value, so combining porch-related features provides a better representation of outdoor living space.

---

### Garage Age
GarageAge = YrSold − GarageYrBlt

This measures how old the garage is. Older garages may negatively affect property value.

---

### Total Living Area
TotalLivingArea = GrLivArea + TotalBsmtSF

This captures total usable indoor living space.

---

## 4. Data Preprocessing

After feature engineering, the following preprocessing steps were applied.

### Removing Sparse Columns
Columns with a very high number of missing values were removed:

- Alley
- PoolQC
- Fence
- MiscFeature

These columns contained very little usable data and could introduce noise.

---

### Handling Missing Values

#### Categorical Features
Missing categorical values were replaced with **"None"** to represent the absence of a feature.

Example:
A missing value in FireplaceQu means the house does not have a fireplace.

---

#### Numerical Features
Missing numerical values were filled using the **median** of each column.

Median was chosen instead of mean because housing data often contains **outliers**, and median is more robust to extreme values.

---

### Encoding Categorical Variables
Categorical features were converted into numerical format using **one-hot encoding**.

This was implemented using:

pd.get_dummies()

One-hot encoding creates binary columns for each category so that machine learning models can interpret categorical data.

---

### Feature Scaling

Feature scaling was applied using **StandardScaler**.

StandardScaler transforms features so that they have:

Mean = 0  
Standard Deviation = 1

#### Why StandardScaler was chosen

StandardScaler is suitable for models that rely on feature magnitude or distance calculations.

Models that benefit from scaling include:

- Linear Regression
- Support Vector Regression (SVR)

StandardScaler was chosen instead of other scalers because:

- It works well when features follow roughly normal distributions.
- It preserves relative relationships between feature values.
- It is widely used for linear models and SVM-based models.

Other scaling methods include:

**MinMaxScaler**
- Scales values between 0 and 1
- Sensitive to outliers

**RobustScaler**
- Uses median and interquartile range
- Better for extremely skewed datasets

For this experiment, **StandardScaler provided a balanced and commonly used scaling approach**.

---

## 5. Model Training

The dataset was split into:

- **80% training data**
- **20% testing data**

The following regression models were trained and evaluated:

- Linear Regression
- Decision Tree Regressor
- Random Forest Regressor
- Gradient Boosting Regressor
- Support Vector Regressor (SVR)

---

## 6. Evaluation Metrics

Three evaluation metrics were used.

### Mean Squared Error (MSE)
Measures the average squared difference between predicted and actual values.

Lower values indicate better performance.

---

### Root Mean Squared Error (RMSE)
The square root of MSE.

It represents the average prediction error in the same units as the target variable.

---

### R² Score
R² measures how well the model explains the variance in the target variable.

Interpretation:

- R² = 1 → perfect prediction
- R² = 0 → model predicts the mean
- R² < 0 → model performs worse than predicting the mean

---

## 7. Results

| Model | MSE | RMSE | R² |
|------|------|------|------|
| Linear Regression | 8.11e+08 | 28486 | 0.894 |
| Decision Tree | 1.63e+09 | 40429 | 0.787 |
| Random Forest | 8.59e+08 | 29320 | 0.888 |
| Gradient Boosting | 6.83e+08 | 26141 | **0.911** |
| SVR | 7.86e+09 | 88646 | -0.024 |

---

## 8. Observations

- **Gradient Boosting achieved the best performance** with an R² score of approximately **0.91**.
- **Linear Regression improved slightly**, indicating that engineered features helped strengthen linear relationships.
- **Random Forest remained stable** and performed well across experiments.
- **Decision Tree performed worse than ensemble models**, likely due to overfitting.
- **SVR continued to perform poorly**, even after scaling and feature engineering.

This suggests that SVR may require **hyperparameter tuning** or may not be well suited for this dataset.

---

## 9. Conclusion

Feature engineering improved the ability of the models to capture important housing characteristics such as size, age, and bathroom capacity.

Among all models, **Gradient Boosting performed the best**, demonstrating its strength in handling complex relationships in tabular data.

Future experiments will focus on:

- Advanced encoding strategies
- Feature selection techniques
- Hyperparameter tuning